### DISCUSSION TITLE
Why not reference python3 as just python too when installed from brew?

### DISCUSSION THREAD

**Discussion by programmer04** [Homebrew/discussions#3596] [Everyday usage]
Python has been removed from macOS default installation - check [macOS Monterey 12.3 Release Notes](https://developer.apple.com/documentation/macos-release-notes/macos-12_3-release-notes#Python). Python in version 2 reaches the end of life over two years ago at January 1, 2020. It shouldn't be used anywhere.

Thus my question is why not just install (link) the newest available `python3` in [homebrew-core](https://github.com/Homebrew/homebrew-core/blob/HEAD/Formula/python@3.10.rb) as `python`? The same for having `pip3` to be available as `pip`. I think it's more convenient and natural to just type `python` to use it. Linux distributions do it this way e.g. [Fedora](https://developer.fedoraproject.org/tech/languages/python/python-installation.html).

I'm creating this as a GH discussion, not a GH issue because maybe I'm missing some crucial context.  Please let me know what you think

**Comment by carlocab** [comment]
I suggested this to @Homebrew/core a while back, and am still in favour of this (at least on macOS versions that do not include a `/usr/bin/python`).

I think there were reservations about breaking scripts that expect to find Python2 when they call `python`.

That said, those scripts are already broken. Maybe the error about `python` being missing would be more informative than the error you get when you try to run Python2 code in Python3.

**Reply by Bo98** [reply]
> I think there were reservations about breaking scripts that expect to find Python2 when they call `python`.

It's not too rare of a scenario this. Ubuntu has a separate `python-is-python3`, for example.

I don't hold any particular opinion either way though. I'm fine with changing this if there's interest.

**Reply by programmer04** [reply]
Yeah, they have it like in [Debian](https://packages.debian.org/unstable/python-is-python3) which always tries to be extremely backward compatible. Maybe this approach is something to consider...

But on the other hand, [Arch](https://archlinux.org/packages/core/x86_64/python/) does the same as [Fedora](https://developer.fedoraproject.org/tech/languages/python/python-installation.html). I'm in the camp that we should kill with fire any references to `python2` anywhere to finish this painful transition from `2` to `3`

**Reply by SMillerDev** [reply]
The python project actually has rules about this: https://peps.python.org/pep-0394/

**Reply by carlocab** [reply]
> The python project actually has rules about this: https://peps.python.org/pep-0394/

This came up in the last discussion too. The only requirement relevant is that `python` better be what you get when you call `python3`, which is what is suggested here.

**Reply by programmer04** [reply]
Yeah @carlocab and [PEP 394](https://peps.python.org/pep-0394/) was created almost **ten years** before EOL for Python 2 and [the last modifications for allowing more flexibility](https://github.com/python/peps/commit/ae932bd6fd2c493d7d64ce328a5293bac4c48884) were in 2019

**Comment by programmer04** [comment]
To sum up the discussion in the suggested answer started by @carlocab there are two options on the table
- link `python3` to `python` and `pip3` to `pip` on  [macOS Monterey 12.3 and newer](https://developer.apple.com/documentation/macos-release-notes/macos-12_3-release-notes#Python) during installation by default (like e.g. on [Arch](https://archlinux.org/packages/core/x86_64/python/) or [Fedora](https://developer.fedoraproject.org/tech/languages/python/python-installation.html))
- introduce to [homebrew-core](https://github.com/Homebrew/homebrew-core) the new package `python-is-python3` that purpose is only to link `python3` to `python` and `pip3` to `pip` (like e.g. on [Debian](https://packages.debian.org/unstable/python-is-python3)), anybody who wants to have this behavior can just install it

Maybe we can have some discussion/voting here? Or should I move it to a GH issue for better visibility and more opinions?

**Reply by fxcoudert** [reply]
We have discussed it at length previously. Introducing `python3 -> python` would be only on macOS versions where `python` is not provided, and probably not on linux, so it would create some inconsistency, which is not particularly desirable.

On the other hand, the only benefit of having `python3 -> python` listed so far is _“it's more convenient and natural”_, which is subjective. It's shorter by one character, but that seems marginal to me.

**Reply by Bo98** [reply]
> introduce to [homebrew-core](https://github.com/Homebrew/homebrew-core) the new package python-is-python3 that purpose is only to link python3 to python and pip3 to pip (like e.g. on [Debian](https://packages.debian.org/unstable/python-is-python3)), anybody who wants to have this behavior can just install it

It's worth noting, since it hasn't been mentioned here, that there is already an opt-in mechanic where you can add `$(brew --prefix)/opt/python/libexec/bin` to your PATH.

**Reply by aristocrates** [reply]
Although it is helpful that the option exists right now to add `$(brew --prefix)/opt/python/libexec/bin` to `$PATH`, it would be great to have the option to manage this through brew itself:
* Managing my shell profile dotfile setup tends to be more work than installing packages with brew, and it also tends to be a lot more fragile.
  * In the context of explaining to other people how to do this (e.g. onboarding, stack exchange questions or similar, etc) `brew install python-is-python3` would be a simpler and more reliable solution than formatting a properly escaped string to echo into the shell profile dotfile while keeping in mind that the desired destination dotfile differs by shell.
* brew manages the contents of `$(brew --prefix)/opt/python/libexec/bin`, and has complete control over whether that path is going to change or stop existing in the future. So it isn't ideal to have parallel references to it hard-coded into PATH set by a different system that brew has no knowledge of.

On the note of consistency, it seems like there is currently close to consensus in other package managers/distros/ecosystems that python = python3 is a standard if not canonical community convention. In modern releases, python 2 does not exist, and either command `python` points at python 3 by default as soon as python 3 is installed, or by default there is no command `python` but if you want to do so, you can make it point at python 3 through the package manager with one install command:
* the "official" python docker image links python to python3 https://github.com/docker-library/python/blob/69d5e5ca5a4323a79668573a8de02dc96fbc7f3b/Dockerfile-linux.template#L294-L301 - I'm aware this image is not maintained by the python developers
* https://packages.debian.org/bullseye/python-is-python3
* https://packages.ubuntu.com/focal/python-is-python3
* https://packages.fedoraproject.org/pkgs/python3.11/python-unversioned-command/

**Reply by aristocrates** [reply]
Case in point regarding fragility of opt-in via adding `$(brew --prefix)/opt/python/libexec/bin` to path: just around 10 days after I posted the above, having set up this workaround on a new machine with a fresh install, it already had broken on its own:

```
$ ls /opt/homebrew/opt/python/libexec/bin
ls: /opt/homebrew/opt/python/libexec/bin: No such file or directory
```

Root cause was of course things ending up in a weird state when switching to and linking python@3.11 and manually fixing that fixed the issue:

```
Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  python@3.11
```

But `$PATH` of course had no idea that python was in a broken state and that `/opt/homebrew/opt/python/libexec/bin` didn't exist, because `$(brew --prefix)/opt/python/libexec/bin` was just hardcoded into shellenv dotfiles. And maybe one day that path stops existing not due to a bug or `brew link` gone wrong, but rather an intentional breaking change in how python and/or brew chooses to package python.

**Reply by MikeMcQuaid** [reply]
@aristocrates Try to use `$(brew --prefix python)` which will error out if Python isn't installed.

**Reply by aristocrates** [reply]
@MikeMcQuaid thanks, I made a PR to change the unversioned symlink note on https://docs.brew.sh/Homebrew-and-Python#python-3y since (at least with a typical setup with bash and `.bash_profile`) you get a visible error message when opening a shell if the package is not installed which is really helpful.

https://github.com/Homebrew/brew/pull/14738

That said, this general class of failure modes still makes a good case for a `python-is-python3` package:

* <details>
  <summary>I'm not sure this would have fixed the particular case I ran into</summary>

  Does `brew --prefix python` error out when the package is installed but not linked (such as when there was an error during `brew link` during package upgrade)? I can't actually reproduce the state I was in before fixing this particular issue so I can't test directly what `brew --prefix python` would have said, but with `brew unlink python@3.11` I still get no error with `brew --prefix python` nor with `brew --prefix python@3.11`
  </details>

* This also doesn't mitigate the possibility that brew one day changes its organization of python executables. If there was a `python-is-python3` package, any breaking changes with executables organization could ship at the same time as any adjustments needed on the `python-is-python3` package's end.

**Reply by MikeMcQuaid** [reply]
Thanks for the PR @aristocrates, queued for merge now.



> * but with `brew unlink python@3.11` I still get no error with `brew --prefix python` nor with `brew --prefix python@3.11`

the `opt` link is resistant to `brew unlink`

> * If there was a `python-is-python3` package, any breaking changes with executables organization could ship at the same time as any adjustments needed on the `python-is-python3` package's end.

This is something we'll consider 👍🏻
